### PR TITLE
AAInstr: add more core functionality

### PR DESCRIPTION
**Pull request by webmiche**
This PR adds more core functionality that will be useful for specific kinds of Instrumentations. In particular, it adds measuring, assembling, performing a single instrumentation run and getting the candidate counts.

### CODE REVIEW

File: AAInstrumentation/core.py [hunk 1]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
             stderr=DEVNULL,
         )
 
+    def get_baseline_files(self) -> list[str]:
+        """Get the baseline files."""
+
+        files = [
+            Path(os.path.join(dirpath.removeprefix(str(self.initial_dir) + "/"), f))
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: Would this not be simpler with `pathlib` instead of `os.path`?

File: AAInstrumentation/core.py [hunk 2]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+            for (dirpath, dirnames, filenames) in os.walk(
+                self.initial_dir.joinpath(self.benchmark)
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: `self.initial_dir/self.benchmark` ?

File: AAInstrumentation/core.py [hunk 3]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+        line_list = second_line.split("\t")
+        # The other results are more info on where size is: text, data, bss
+        return int(line_list[0])
+
+    def assemble_file(self, file_name: Path, obj_file_name: Path):
+        """Assemble the file."""
+
+        cmd = [
+            str(self.instr_path.joinpath("llc")),
+            "-O2",
```
Comment (thetheodor) [inline AAInstrumentation/core.py:98]: Would it not make sense to not hardcode this?

File: AAInstrumentation/core.py [hunk 4]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+        second_line = p.stdout.split("\n")[1]
+        line_list = second_line.split("\t")
+        # The other results are more info on where size is: text, data, bss
+        return int(line_list[0])
+
+    def assemble_file(self, file_name: Path, obj_file_name: Path):
+        """Assemble the file."""
+
+        cmd = [
+            str(self.instr_path.joinpath("llc")),
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: `str(self.instr_path/"llc")` ?

File: AAInstrumentation/core.py (hunk 3 above)
Comment (webmiche) [reply AAInstrumentation/core.py:98]: This is just for llc and I think this is what's being run for O3 and Oz, so I decided to hardcode it here. I guess I can also make it a class variable. What do you think?

File: AAInstrumentation/core.py [hunk 5]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+        obj_file_name = file_name.with_suffix(".o")
+        self.assemble_file(file_name, obj_file_name)
+
+        return self.measure_outputsize(obj_file_name)
+
+    def run_step_single_func(
+        self,
+        file_name: Path,
+        index: int,
+        index_list: list[int],
```
Comment (thetheodor) [inline AAInstrumentation/core.py:120]: maybe add some comments describing what `index` and `index_list` are?

File: AAInstrumentation/core.py [hunk 6]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+    ):
+        """Run the step with the given file name."""
+
+        cmd = [
+            str(self.instr_path.joinpath("opt")),
+            str(self.initial_dir.joinpath(file_name)),
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: `/` instead of `joinpath`?

File: AAInstrumentation/core.py [hunk 7]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+            "-stats",
+            "-o",
+            str(
+                self.instr_dir.joinpath(
+                    file_name.parent, str(index) + str(file_name.stem) + ".bc"
+                )
+            ),
```
Comment (thetheodor) [inline AAInstrumentation/core.py:139]: `str(self.instr_dir/file_name.parent/str(index)+ file_name.stem+".bc")` ?

File: AAInstrumentation/core.py [hunk 8]
```diff
@@ -67,6 +70,119 @@ def generate_baseline(self):
...
+            "-" + self.opt_flag,
+            "--aasequence=0-",
+            "-o",
+            "/dev/null",
+        ]
+        p = run(
+            cmd,
+            cwd=self.exec_root,
+            stderr=PIPE,
+            text=True,
```
Comment (thetheodor) [inline AAInstrumentation/core.py:171]: would it make sense to add a 'check_output=True` to abort on failures?

File: AAInstrumentation/core.py [hunk 9]
```diff
@@ -107,12 +223,25 @@ def generate_baseline(self):
...
 
-    AAInstrumentationDriver(
+    driver = AAInstrumentationDriver(
         instr_path,
         exec_root,
         specbuild_dir,
         benchmark,
         initial_dir,
+        instr_dir,
         "Oz",
```
Comment (thetheodor) [inline AAInstrumentation/core.py:239]: does this `Oz` correspond to the `O2` of  the `llc` call?

File: AAInstrumentation/core.py [hunk 10]
```diff
@@ -107,12 +223,25 @@ def generate_baseline(self):
...
+    driver.generate_baseline()
+
+    files = driver.get_baseline_files()
+
+    count_per_file = driver.get_candidates_per_file(files)
+
+    print(count_per_file)
+    for file in files:
+        driver.run_step_single_func(file, 0, [1, 2, 3])
+        file_name = initial_dir.joinpath(file)
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: `initial_dir/file` :P

**Review by thetheodor — COMMENTED** [review]
Not sure if you want to address these. Many are nitpicks

**Review by thetheodor — APPROVED** [review]

**Pull request merged**